linked_list/linked_list.py

```diff
@@ -13,67 +15,125 @@ def __init__(self):
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
+        if self.head == None:
+           return None
+        return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity:  O(1)
+    # Space Complexity:  O(1)
     def add_first(self, value):
-        pass
+        new_node = Node(value)
+        new_node.next = self.head
+        self.head = new_node   
+        #=> DRY VERSION 
+        # self.head = Node(value, self.head)
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity:  O(N)
+    # Space Complexity:  O(N)
     def search(self, value):
-        pass
+        current = self.head
+        while current != None:
+            if current.value == value:
+                return True
+            current = current.next
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(N)
+    # Space Complexity: O(N)
     def length(self):
-        pass
+        current = self.head  #initialize current 
+        counter = 0
+        
+        while current:  #loop until current is None 
+            counter += 1
+            current = current.next
+        return counter
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity:  O(N)
+    # Space Complexity:  O(N)
     def get_at_index(self, index):
-        pass
+        if index < 0:
+            return None
+        current = self.head  
+        counter = 0
+        while current:
+            if counter == index:
+                return current.value
+            counter +=1
+            current = current.next #goes to next node
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity:  O(N)
+    # Space Complexity:  O(N)
     def get_last(self):

```

Comment:
👍   How is this O(n) space complexity?

Also note if you maintain a `tail` attribute you can do this method in O(1) time complexity